File: backend/src/repos/scan.rs
//! Scan repos for contribution data.

use git2::{ErrorCode, Oid, Repository};
use std::path::Path;

// FIXME use super::Error and super::Result

/// Scan history of a repository and commit dates as seconds since 1970.
///
/// The path must be one of:
///
///   * A repository working directory containing a `.git` directory
///   * A `.git` directory itself
///   * A bare repository
///
/// # Errors
///
/// Returns an error if there was a problem with the repository. Returns
/// `Ok(None)` if the remote HEAD could not be found.
pub fn scan_repo_path<P: AsRef<Path>>(path: P) -> anyhow::Result<Vec<i64>> {
    scan_repo(&Repository::open(path)?)
}

/// Scan history of a repository and commit dates as seconds since 1970.
///
/// # Errors
///
/// Returns an error if there was a problem with the repository. Returns
/// `Ok(None)` if the remote HEAD could not be found.
pub fn scan_repo(repo: &Repository) -> anyhow::Result<Vec<i64>> {
    let mut revwalk = repo.revwalk()?;
    revwalk.set_sorting(git2::Sort::TIME)?;

    let default_branch_oid = get_default_branch(repo)?;
    revwalk.push(default_branch_oid)?;

    for remote_name in repo.remotes()?.into_iter().flatten().flatten() {
        let remote = repo.find_remote(remote_name)?;
        if let Ok(url) = remote.url()
            && (url.starts_with("git@github.com:")
                || url.starts_with("https://github.com/"))
        {
            // GitHub remote. Skip; any local commits are equivalent to branch
            // commits on GitHub.
            return Ok(Vec::new());
        }
        // FIXME warn about non UTF-8?
    }

    revwalk
        .map(|oid| {
            oid.and_then(|oid| repo.find_commit(oid))
                .map(|commit| commit.author().when().seconds())
                .map_err(anyhow::Error::from)
        })
        .collect()
}

/// Find the default branch of a repository.
///
/// `git` doesn’t really have a concept of a default branch, so this involves
/// some guess work. We check:
///
///   1. `refs/remotes/origin/HEAD` to see if it points to a remote branch
///   2. `refs/remotes/upstream/HEAD`?
///   3. Check if `$(git config init.defaultBranch)` is a branch
///   4. Check if `main` is a branch.
///   5. Check if `master` is a branch.
///   6. Return `HEAD`
///
/// # Errors
///
/// Returns an error if there was a problem with the repository. Returns
/// `Ok(None)` if the remote HEAD could not be found.
pub fn get_default_branch(repo: &Repository) -> anyhow::Result<Oid> {
    if let Some(branch) = remote_head_to_local_branch(repo, "origin")?
        && let Some(oid) = ref_to_oid(repo, &branch)?
    {
        return Ok(oid);
    }

    if let Some(branch) = remote_head_to_local_branch(repo, "upstream")?
        && let Some(oid) = ref_to_oid(repo, &branch)?
    {
        return Ok(oid);
    }

    match repo.config()?.get_string("init.defaultBranch") {
        Ok(branch) => {
            if let Some(oid) = ref_to_oid(repo, &branch)? {
                return Ok(oid);
            }
        }
        Err(error) if error.code() == ErrorCode::NotFound => {}
        Err(error) => return Err(error.into()),
    }

    if let Some(oid) = ref_to_oid(repo, "refs/heads/main")? {
        Ok(oid)
    } else if let Some(oid) = ref_to_oid(repo, "refs/heads/master")? {
        Ok(oid)
    } else if let Some(oid) = ref_to_oid(repo, "HEAD")? {
        Ok(oid)
    } else {
        Err(anyhow::anyhow!("Could not find a default branch"))
    }
}

/// Get the branch name a remote HEAD points to.
///
/// # Errors
///
/// Returns an error if there was a problem with the repository. Returns
/// `Ok(None)` if the remote HEAD could not be found.
fn remote_head_to_local_branch(
    repo: &Repository,
    origin: &str,
) -> anyhow::Result<Option<String>> {
    match repo.find_reference(&format!("refs/remotes/{origin}/HEAD")) {
        Ok(reference) => {
            if let Ok(Some(target)) = reference.symbolic_target()
                && target.starts_with("refs/remotes/")
            {
                let mut iter = target.splitn(4, '/');
                if let Some(branch) = iter.nth(3) {
                    assert!(iter.next().is_none(), "bug in splitn");
                    return Ok(Some(branch.to_owned()));
                }
            }
            // FIXME? log failure?
            // else: try next thing, though it’s weird this isn’t symbolic.
            // FIXME log?
            Ok(None)
        }
        Err(error) if error.code() == ErrorCode::NotFound => Ok(None),
        Err(error) => Err(error.into()),
    }
}

/// Find the `Oid` for a ref.
///
/// # Errors
///
/// Returns an error if there was a problem parsing `name` or with the
/// repository. Returns `Ok(None)` if the ref could not be found.
fn ref_to_oid(repo: &Repository, name: &str) -> anyhow::Result<Option<Oid>> {
    match repo.revparse_single(name) {
        Ok(object) => Ok(Some(object.id())),
        Err(error) if error.code() == ErrorCode::NotFound => Ok(None),
        Err(error) => Err(error.into()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test::{FsDirectory, Home};
    use assert2::assert;
    use testdir::testdir;

    #[test]
    fn default_branch_prefers_origin_head_over_upstream_head() {
        let home = Home::init(testdir!());
        let repo = home.git_init("repo");
        repo.make_commit(0); // main: 1 commit (c0)

        repo.git(["branch", "upstream-branch"]);
        repo.git(["checkout", "upstream-branch"]);
        repo.make_commit(1); // upstream-branch: 2 commits (c0, c1)

        repo.git(["branch", "origin-branch"]); // branched from upstream-branch
        repo.git(["checkout", "origin-branch"]);
        repo.make_commit(2); // origin-branch: 3 commits (c0, c1, c2)

        repo.git(["checkout", "main"]);

        repo.git([
            "symbolic-ref",
            "refs/remotes/upstream/HEAD",
            "refs/remotes/upstream/upstream-branch",
        ]);
        repo.git([
            "symbolic-ref",
            "refs/remotes/origin/HEAD",
            "refs/remotes/origin/origin-branch",
        ]);

        // Both origin/HEAD and upstream/HEAD are set, pointing at different
        // (real) branches with different history lengths: origin must win.
        assert!(let Ok([_, _, _]) = scan_repo_path(repo.path()).as_deref());
    }

    #[test]
    fn default_branch_falls_back_to_upstream_head_over_config() {
        let home = Home::init(testdir!());
        let repo = home.git_init("repo");
        repo.make_commit(0); // main: 1 commit (c0)

        repo.git(["branch", "config-branch"]);
        repo.git(["checkout", "config-branch"]);
        repo.make_commit(1); // config-branch: 2 commits (c0, c1)

        repo.git(["branch", "upstream-branch"]); // branched from config-branch
        repo.git(["checkout", "upstream-branch"]);
        repo.make_commit(2); // upstream-branch: 3 commits (c0, c1, c2)

        repo.git(["checkout", "main"]);

        repo.git(["config", "init.defaultBranch", "config-branch"]);
        repo.git([
            "symbolic-ref",
            "refs/remotes/upstream/HEAD",
            "refs/remotes/upstream/upstream-branch",
        ]);

        // No origin/HEAD; upstream/HEAD and init.defaultBranch both point at
        // real, different branches: upstream must win over the config.
        assert!(let Ok([_, _, _]) = scan_repo_path(repo.path()).as_deref());
    }

    #[test]
    fn default_branch_uses_local_config_when_no_remote_head() {
        let home = Home::init(testdir!());
        let repo = home.git_init("repo");
        repo.make_commit(0); // main: 1 commit

        repo.git(["branch", "feature"]);
        repo.git(["checkout", "feature"]);
        repo.make_commit(1); // feature: 2 commits
        repo.git(["checkout", "main"]);

        repo.git(["config", "init.defaultBranch", "feature"]);

        assert!(let Ok([_, _]) = scan_repo_path(repo.path()).as_deref());
    }

    #[test]
    fn default_branch_local_config_falls_through_when_branch_missing() {
        let home = Home::init(testdir!());
        let repo = home.git_init("repo");
        repo.make_commit(0); // main: 1 commit

        // "ghost" doesn't exist, so this should fall through to `main`
        // rather than erroring.
        repo.git(["config", "init.defaultBranch", "ghost"]);

        assert!(let Ok([_]) = scan_repo_path(repo.path()).as_deref());
    }

    #[test]
    fn default_branch_uses_master_when_no_main() {
        let home = Home::init(testdir!());
        let repo = home.git_init("repo");
        repo.make_commit(0);
        repo.git(["branch", "-m", "main", "master"]);

        // The home's global `init.defaultBranch = main` config still points
        // at a branch that no longer exists, so this exercises the
        // `master` fallback specifically.
        assert!(let Ok([_]) = scan_repo_path(repo.path()).as_deref());
    }

    #[test]
    fn default_branch_falls_back_to_head() {
        let home = Home::init(testdir!());
        let repo = home.git_init("repo");
        repo.make_commit(0);
        // No `main`, no `master`, and the global `init.defaultBranch = main`
        // config points at a branch that doesn't exist either.
        repo.git(["branch", "-m", "main", "custom"]);

        assert!(let Ok([_]) = scan_repo_path(repo.path()).as_deref());
    }

    #[test]
    fn scan_empty_repo_errors() {
        let home = Home::init(testdir!());
        let repo = home.git_init("repo");
        // No commits: HEAD is unborn, and there's no branch to fall back to.
        assert!(let Err(_) = scan_repo_path(repo.path()).as_deref());
    }

    #[test]
    fn scan_repo_uses_author_time_not_committer_time() {
        let home = Home::init(testdir!());
        let repo = home.git_init("repo");
        repo.write("a", "a");
        repo.git(["add", "a"]);
        repo.git([
            "commit",
            "-m",
            "commit",
            "--date",
            "2000-01-01T00:00:00+00:00",
        ]);

        // If this used committer time instead of author time, it would be
        // ~now, not the fixed date above (946_684_800 = 2000-01-01 UTC).
        assert!(scan_repo_path(repo.path()).unwrap() == vec![946_684_800]);
    }

    #[test]
    fn scan_repo_skips_when_origin_is_github_ssh() {
        let home = Home::init(testdir!());
        let repo = home.git_init("repo");
        repo.make_commit(0);
        repo.git([
            "remote",
            "add",
            "origin",
            "git@github.com:example/repo.git",
        ]);

        assert!(scan_repo_path(repo.path()).unwrap().is_empty());
    }

    #[test]
    fn scan_repo_skips_when_origin_is_github_https() {
        let home = Home::init(testdir!());
        let repo = home.git_init("repo");
        repo.make_commit(0);
        repo.git([
            "remote",
            "add",
            "origin",
            "https://github.com/example/repo.git",
        ]);

        assert!(scan_repo_path(repo.path()).unwrap().is_empty());
    }

    #[test]
    fn scan_repo_returns_history_for_non_github_remote() {
        let home = Home::init(testdir!());
        let repo = home.git_init("repo");
        repo.make_commit(0);
        repo.git([
            "remote",
            "add",
            "origin",
            "https://git.example.com/repo.git",
        ]);

        assert!(let Ok([_]) = scan_repo_path(repo.path()).as_deref());
    }

    #[test]
    fn scan_repo_skips_if_any_remote_is_github() {
        let home = Home::init(testdir!());
        let repo = home.git_init("repo");
        repo.make_commit(0);
        repo.git([
            "remote",
            "add",
            "origin",
            "https://git.example.com/repo.git",
        ]);
        repo.git([
            "remote",
            "add",
            "github",
            "https://github.com/example/repo.git",
        ]);

        // Documents current behavior: a single GitHub remote is enough to
        // skip the whole repo, even if other remotes aren't GitHub. Worth
        // confirming this is actually intended.
        assert!(scan_repo_path(repo.path()).unwrap().is_empty());
    }

    #[test]
    fn scan_repo() {
        let home = Home::init(testdir!());
        let repo = home.git_init("repo");
        repo.make_commit(0);
        assert!(let Ok([_]) = scan_repo_path(repo.path()).as_deref());
    }

    #[test]
    fn scan_repo_dotgit() {
        let home = Home::init(testdir!());
        let repo = home.git_init("repo");
        repo.make_commit(0);
        assert!(let Ok([_]) = scan_repo_path(repo.path().join(".git")).as_deref());
    }

    #[test]
    fn scan_repo_subdir() {
        let home = Home::init(testdir!());
        let repo = home.git_init("repo");

        repo.write("dir/a", "a0");
        repo.git(["add", "dir/a"]);
        repo.git(["commit", "-m", "commit 0"]);

        // FIXME check error code.
        assert!(let Err(_) = scan_repo_path(repo.join("dir")).as_deref());
    }

    #[test]
    fn scan_nonrepo() {
        let home = Home::init(testdir!());
        let repo = home.git_init("repo");
        repo.make_commit(0);

        // FIXME check error code.
        assert!(let Err(_) = scan_repo_path(home.path()).as_deref());
    }

    #[test]
    fn scan_bare_repo() {
        let home = Home::init(testdir!());
        let bare_repo = home.git_init_bare("bare_repo");
        let repo = bare_repo.clone("repo");
        repo.make_commit(0);
        repo.git(["push"]);

        assert!(let Ok([_]) = scan_repo_path(bare_repo.path()).as_deref());
    }
}

File: backend-ts/src/repos/scan.ts
// Scan repos for contribution data.

import { isAbsolute, join } from "@std/path";

/**
 * Resolve `path` to the git directory it names, matching `git2`'s
 * `Repository::open` semantics: it must be one of
 *
 *   - A working directory containing a `.git` directory (or gitlink file)
 *   - A `.git` directory itself
 *   - A bare repository
 *
 * Unlike `git -C <path>` (or `git rev-parse --show-toplevel`), this does
 * NOT discover a repository in a parent directory — a path that isn't
 * itself one of the above returns `null`.
 */
export async function resolveGitDir(path: string): Promise<string | null> {
  const dotGit = join(path, ".git");
  try {
    const stat = await Deno.lstat(dotGit);
    if (stat.isDirectory) {
      return (await isGitDirLayout(dotGit))
        ? await Deno.realPath(dotGit)
        : null;
    }
    if (stat.isFile) {
      // Worktree/submodule gitlink file: `gitdir: <path>`.
      const content = await Deno.readTextFile(dotGit);
      const match = content.match(/^gitdir:\s*(.+)$/m);
      if (!match) return null;
      const target = match[1].trim();
      const resolved = isAbsolute(target) ? target : join(path, target);
      return (await isGitDirLayout(resolved))
        ? await Deno.realPath(resolved)
        : null;
    }
    return null;
  } catch {
    // No `.git` entry; fall through to check whether `path` itself is a
    // git directory (a bare repo, or a `.git` directory passed directly).
  }
  return (await isGitDirLayout(path)) ? await Deno.realPath(path) : null;
}

async function isGitDirLayout(dir: string): Promise<boolean> {
  try {
    const [head, objects, refs] = await Promise.all([
      Deno.stat(join(dir, "HEAD")),
      Deno.stat(join(dir, "objects")),
      Deno.stat(join(dir, "refs")),
    ]);
    return head.isFile && objects.isDirectory && refs.isDirectory;
  } catch {
    return false;
  }
}

interface GitResult {
  success: boolean;
  stdout: string;
  stderr: string;
}

/** Run `git --git-dir=<gitDir> ...args`, pinned to an exact git directory. */
async function runGit(gitDir: string, args: string[]): Promise<GitResult> {
  const command = new Deno.Command("git", {
    args: ["--git-dir", gitDir, ...args],
    stdout: "piped",
    stderr: "piped",
  });
  const { success, stdout, stderr } = await command.output();
  return {
    success,
    stdout: new TextDecoder().decode(stdout),
    stderr: new TextDecoder().decode(stderr),
  };
}

/**
 * Split `s` on `sep` into at most `n` pieces, with the final piece holding
 * any remainder (unlike `String.prototype.split`, which truncates it).
 */
function splitN(s: string, sep: string, n: number): string[] {
  const parts: string[] = [];
  let rest = s;
  for (let i = 0; i < n - 1; i++) {
    const idx = rest.indexOf(sep);
    if (idx === -1) break;
    parts.push(rest.slice(0, idx));
    rest = rest.slice(idx + sep.length);
  }
  parts.push(rest);
  return parts;
}

/** Resolve `name` (a ref, branch name, or `HEAD`) to a commit oid. */
async function refToOid(
  gitDir: string,
  name: string,
): Promise<string | null> {
  const result = await runGit(gitDir, [
    "rev-parse",
    "--verify",
    "--quiet",
    name,
  ]);
  if (!result.success) return null;
  const oid = result.stdout.trim();
  return oid.length > 0 ? oid : null;
}

/** Get the local branch name a remote `HEAD` points to, if any. */
async function remoteHeadToLocalBranch(
  gitDir: string,
  remote: string,
): Promise<string | null> {
  const result = await runGit(gitDir, [
    "symbolic-ref",
    `refs/remotes/${remote}/HEAD`,
  ]);
  if (!result.success) return null;

  const target = result.stdout.trim();
  if (!target.startsWith("refs/remotes/")) return null;

  const parts = splitN(target, "/", 4);
  return parts.length === 4 ? parts[3] : null;
}

/**
 * Find the default branch of a repository, as a commit oid.
 *
 * `git` doesn't really have a concept of a default branch, so this involves
 * some guesswork. We check:
 *
 *   1. `refs/remotes/origin/HEAD` to see if it points to a remote branch
 *   2. `refs/remotes/upstream/HEAD`
 *   3. Check if `$(git config init.defaultBranch)` is a branch
 *   4. Check if `main` is a branch
 *   5. Check if `master` is a branch
 *   6. Return `HEAD`
 *
 * @throws if none of the above resolve to a commit.
 */
export async function getDefaultBranch(gitDir: string): Promise<string> {
  for (const remote of ["origin", "upstream"]) {
    const branch = await remoteHeadToLocalBranch(gitDir, remote);
    if (branch !== null) {
      const oid = await refToOid(gitDir, branch);
      if (oid !== null) return oid;
    }
  }

  const configResult = await runGit(gitDir, [
    "config",
    "--get",
    "init.defaultBranch",
  ]);
  if (configResult.success) {
    const branch = configResult.stdout.trim();
    if (branch.length > 0) {
      const oid = await refToOid(gitDir, branch);
      if (oid !== null) return oid;
    }
  }

  for (const ref of ["refs/heads/main", "refs/heads/master", "HEAD"]) {
    const oid = await refToOid(gitDir, ref);
    if (oid !== null) return oid;
  }

  throw new Error("Could not find a default branch");
}

/** Scan history of a repository; commit dates as seconds since 1970. */
export async function scanRepo(gitDir: string): Promise<number[]> {
  const defaultOid = await getDefaultBranch(gitDir);

  const remotesResult = await runGit(gitDir, ["remote"]);
  const remoteNames = remotesResult.stdout
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0);

  for (const name of remoteNames) {
    const urlResult = await runGit(gitDir, ["remote", "get-url", name]);
    if (!urlResult.success) continue; // FIXME? warn about unreadable remote?
    const url = urlResult.stdout.trim();
    if (
      url.startsWith("git@github.com:") ||
      url.startsWith("https://github.com/")
    ) {
      // GitHub remote. Skip; any local commits are equivalent to branch
      // commits on GitHub.
      return [];
    }
  }

  const logResult = await runGit(gitDir, ["log", "--format=%at", defaultOid]);
  if (!logResult.success) {
    throw new Error(`git log failed: ${logResult.stderr}`);
  }
  return logResult.stdout
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0)
    .map((line) => Number.parseInt(line, 10));
}

/** Scan history of a repository at `path`; commit dates as seconds since 1970. */
export async function scanRepoPath(path: string): Promise<number[]> {
  const gitDir = await resolveGitDir(path);
  if (gitDir === null) {
    throw new Error(`Not a git repository: ${path}`);
  }
  return scanRepo(gitDir);
}
